File: spark-examples/Scala/S3/Job2S3Write.scala
import org.apache.spark.sql.{SparkSession, functions => F}

object Job2S3Write {
  def main(args: Array[String]): Unit = {

    // ==========================
    // Environment Variables (Driver)
    // ==========================
    val accessKey = sys.env.getOrElse("DATASTORE_AWS_ACCESS_KEY_ID", "")
    val secretKey = sys.env.getOrElse("DATASTORE_AWS_SECRET_ACCESS_KEY", "")
    val bucketName = sys.env.getOrElse("DATASTORE_S3_BUCKET_NAME", "")
    val s3FilePath = sys.env.getOrElse("DATASTORE_S3_FILE_PATH", "")
    val s3FilePathOutput = sys.env.getOrElse("DATASTORE_S3_FILE_PATH_OUTPUT", "")
    val s3Region = sys.env.getOrElse("DATASTORE_S3_REGION", "")
    
    
    // ==========================
    // Initialize Spark Session
    // ==========================
    val spark = SparkSession.builder()
      .appName("S3WriteApplication")
      .getOrCreate()

    // ==========================
    // Configure S3 credentials in Hadoop configuration
    // ==========================
    val hadoopConf = spark.sparkContext.hadoopConfiguration
    hadoopConf.set("fs.s3a.access.key", accessKey)
    hadoopConf.set("fs.s3a.secret.key", secretKey)
    hadoopConf.set("fs.s3a.endpoint.region", s3Region)
    
    // Optional: specify S3 endpoint if using non-AWS S3
    // hadoopConf.set("fs.s3a.endpoint", "s3.amazonaws.com")

    // ==========================
    // Build S3 File Path
    // ==========================
    val filePath = s"s3a://$bucketName/$s3FilePath"

    // ==========================
    // Read Data from S3
    // ==========================
    val df = spark.read
      .option("header", "true")
      .option("inferSchema", "true")
      .csv(filePath)

    // ==========================
    // Show the data
    // ==========================
    df.show(truncate = false)

    // ==========================
    // Build Output File Path
    // ==========================
    val outPutFilePath = s"s3a://$bucketName/$s3FilePathOutput"

    // ==========================
    // Write Data to S3
    // ==========================
    df.write
      .option("header", "true")
      .mode("overwrite")
      .csv(outPutFilePath)

    // ==========================
    // Read Data from S3
    // ==========================
    val df_output = spark.read
      .option("header", "true")
      .option("inferSchema", "true")
      .csv(outPutFilePath)

    // ==========================
    // Show the data
    // ==========================
    df_output.show(truncate = false)

    spark.stop()
  }
}

File: spark-examples/Java/S3/Job2S3Write.java
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.hadoop.conf.Configuration;

public class Job2S3Write {
    public static void main(String[] args) {

        // ==========================
        // Environment Variables (Driver)
        // ==========================
        String accessKey = System.getenv().getOrDefault("DATASTORE_AWS_ACCESS_KEY_ID", "");
        String secretKey = System.getenv().getOrDefault("DATASTORE_AWS_SECRET_ACCESS_KEY", "");
        String bucketName = System.getenv().getOrDefault("DATASTORE_S3_BUCKET_NAME", "");
        String s3FilePath = System.getenv().getOrDefault("DATASTORE_S3_FILE_PATH", "");
        String s3FilePathOutput = System.getenv().getOrDefault("DATASTORE_S3_FILE_PATH_OUTPUT", "");
        String s3Region = System.getenv().getOrDefault("DATASTORE_S3_REGION", "");

        // ==========================
        // Initialize Spark Session
        // ==========================
        SparkSession spark = SparkSession.builder()
                .appName("S3WriteApplication")
                .getOrCreate();

        // ==========================
        // Configure S3 credentials in Hadoop configuration
        // ==========================
        Configuration hadoopConf = spark.sparkContext().hadoopConfiguration();
        hadoopConf.set("fs.s3a.access.key", accessKey);
        hadoopConf.set("fs.s3a.secret.key", secretKey);
        hadoopConf.set("fs.s3a.endpoint.region", s3Region);

        // Optional: specify S3 endpoint if using non-AWS S3
        // hadoopConf.set("fs.s3a.endpoint", "s3.amazonaws.com");

        // ==========================
        // Build S3 File Path
        // ==========================
        String filePath = "s3a://" + bucketName + "/" + s3FilePath;

        // ==========================
        // Read Data from S3
        // ==========================
        Dataset<Row> df = spark.read()
                .option("header", "true")
                .option("inferSchema", "true")
                .csv(filePath);

        // ==========================
        // Show the data
        // ==========================
        df.show(false);

        // ==========================
        // Build Output File Path
        // ==========================
        String outPutFilePath = "s3a://" + bucketName + "/" + s3FilePathOutput;

        // ==========================
        // Write Data to S3
        // ==========================
        df.write()
                .option("header", "true")
                .mode("overwrite")
                .csv(outPutFilePath);

        // ==========================
        // Read Data from S3
        // ==========================
        Dataset<Row> dfOutput = spark.read()
                .option("header", "true")
                .option("inferSchema", "true")
                .csv(outPutFilePath);

        // ==========================
        // Show the data
        // ==========================
        dfOutput.show(false);

        spark.stop();
    }
}
